### PR TITLE
Compute intersecting lanes for Waymo maps

### PR DESCRIPTION
**Pull request by saulfield**
- When loading a Waymo map, we now compute a set of intersecting lanes for each lane in the map. This is to be used for identifying junctions and foes for traffic simulation.
- We now cache TFRecord file data to speed up Waymo map loading for cases where multiple scenarios are being loaded consecutively from the same TFRecord file.

### CODE REVIEW

File: smarts/core/waymo_map.py
Comment (sah-huawei) [inline smarts/core/waymo_map.py:311]: I don't think the `np.array()` is necessary here (as `np.roll()` will make an array out of it).

File: smarts/core/waymo_map.py
Comment (sah-huawei) [inline smarts/core/waymo_map.py:317]: Re our discussion about how to avoid the repeated allocations from `np.tile()` here, I think this page has answers:  https://numpy.org/doc/stable/user/basics.broadcasting.html

Long story short, I don't think you need the `tile()` calls here.

File: smarts/core/waymo_map.py
Comment (sah-huawei) [inline smarts/core/waymo_map.py:105]:
```suggestion
    _tfrecord_path: Optional[str] = None
    _tfrecord_generator = Optional[Generator[bytes, None, None]] = None
    _scenario_cache: Optional[Dict[str, Any]] = None
```

File: smarts/core/waymo_map.py
Comment (sah-huawei) [inline smarts/core/waymo_map.py:822]: It's minor here, I know, but this pattern avoids double lookups in dicts:
```suggestion
        parsed_scenario = WaymoMap._scenario_cache.get(scenario_id)
        if parsed_scenario:
            return parsed_scenario
```

File: smarts/core/waymo_map.py
Comment (sah-huawei) [inline smarts/core/waymo_map.py:323]: good!

**Review by sah-huawei — COMMENTED** [review]
I think I can now use `self._intersections` to implment `foes()` on the segmented lanes.  I'm wondering if I should do that on this branch/PR or wait until it's merged and then do a different one.

File: smarts/core/waymo_map.py
Comment (saulfield) [reply smarts/core/waymo_map.py]: Yes you're right -- good catch! It looks like this does indeed give us an additional 10-20% speedup too.

**Review by sah-huawei — APPROVED** [review]

**Review by Gamenot — APPROVED** [review]

**Comment by Gamenot** [comment]
@sah-huawei Sorry, I reviewed this and forgot to actually hit the button.

**Pull request merged**
